planner/core/fragment.go

```diff
@@ -35,56 +37,75 @@ type Fragment struct {
+	rootTasks, err := e.generateMPPTasksForFragment(s.Fragment)
+	if err != nil {
+		return nil, errors.Trace(err)
 	}
-	rootPf.ExchangeSender.InitBasePlan(ctx, plancodec.TypeExchangeSender)
-	rootPf.ExchangeSender.SetChildren(rootPf.p)
-	getPlanFragments(ctx, p, rootPf)
-	return rootPf
+	s.Tasks = []*kv.MPPTask{tidbTask}
+	return rootTasks, nil
 }
 
-// getPlanFragment passes the plan and which fragment the plan belongs to, then walk through the plan recursively.
-// When we found an edge can be cut, we will add exchange operators and construct new fragment.
-func getPlanFragments(ctx sessionctx.Context, p PhysicalPlan, pf *Fragment) {
-	switch x := p.(type) {
-	case *PhysicalTableScan:
-		x.IsGlobalRead = false
-		pf.TableScan = x
-	case *PhysicalBroadCastJoin:
-		// This is a fragment cutter. So we replace broadcast side with a exchangerClient
-		bcChild := x.Children()[x.InnerChildIdx]
-		exchangeSender := &PhysicalExchangeSender{ExchangeType: tipb.ExchangeType_Broadcast}
-		exchangeSender.InitBasePlan(ctx, plancodec.TypeExchangeSender)
-		npf := &Fragment{p: bcChild, ExchangeSender: exchangeSender}
-		exchangeSender.SetChildren(npf.p)
-
-		exchangeReceivers := &PhysicalExchangeReceiver{
-			ChildPf: npf,
+func (e *mppTaskGenerator) generateMPPTasksForFragment(f *Fragment) (tasks []*kv.MPPTask, err error) {
+	for _, r := range f.ExchangeReceivers {
+		r.Tasks, err = e.generateMPPTasksForFragment(r.ChildPf)
+		if err != nil {
+			return nil, errors.Trace(err)
 		}
-		exchangeReceivers.InitBasePlan(ctx, plancodec.TypeExchangeReceiver)
-		x.Children()[x.InnerChildIdx] = exchangeReceivers
-		pf.ExchangeReceivers = append(pf.ExchangeReceivers, exchangeReceivers)
+	}
+	if f.TableScan != nil {
+		tasks, err = e.constructSinglePhysicalTable(context.Background(), f.TableScan.Table.ID, f.TableScan.Ranges)
+	} else {
+		tasks, err = e.constructSinglePhysicalTable(context.Background(), -1, nil)
+	}
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if len(tasks) == 0 {
+		return nil, errors.New("cannot find mpp task")
+	}
+	for _, task := range tasks {
+		logutil.BgLogger().Info("Dispatch mpp task", zap.Uint64("timestamp", task.StartTs), zap.Int64("ID", task.ID), zap.String("address", task.Meta.GetAddress()), zap.String("plan", ToString(f.ExchangeSender)))
+	}
+	for _, r := range f.ExchangeReceivers {
+		s := r.ChildPf.ExchangeSender
+		s.Tasks = tasks
+	}
+	return tasks, nil
+}
 
-		// For the inner side of join, we use a new plan fragment.
-		getPlanFragments(ctx, bcChild, npf)
-		getPlanFragments(ctx, x.Children()[1-x.InnerChildIdx], pf)
-	default:
-		if len(x.Children()) > 0 {
-			getPlanFragments(ctx, x.Children()[0], pf)
-		}
+// single physical table means a table without partitions or a single partition in a partition table.
+func (e *mppTaskGenerator) constructSinglePhysicalTable(ctx context.Context, tableID int64, ranges []*ranger.Range) ([]*kv.MPPTask, error) {
```

Comment:
this function name is misleading, it aims to generate each MPP task, including which contains table scans or not.

Reply:
please rename it as `generateMPPTasksInternal`